core/Repository/NPMRepositoryClient.cs

```diff
@@ -0,0 +1,137 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Abstractions;
+using System.IO.Compression;
+using System.Linq;
+using System.Threading.Tasks;
+using Cmf.CLI.Core.Interfaces;
+using Cmf.CLI.Core.Objects;
+using Cmf.CLI.Core.Services;
+using Cmf.CLI.Utilities;
+
+namespace Cmf.CLI.Core.Repository;
+
+public class NPMRepositoryClient : IRepositoryClient
+{
+    private IFileSystem fileSystem;
+    private Uri registryUrl;
+    private INPMClientEx client;
+    public NPMRepositoryClient(string registryUrl, IFileSystem fileSystem = null, INPMClientEx client = null)
+    {
+        this.registryUrl = new Uri(registryUrl);
+        this.client = client ?? new NPMClient(registryUrl);
+        this.fileSystem = fileSystem ?? new FileSystem();
+    }
+    public async Task<CmfPackageV1> Find(string packageId, string version)
+    {
+        try
+        {
+            var pkg = await client.FetchPackageVersion(packageId.ToLowerInvariant(), version);
+            pkg.Client = this;
+            return pkg;
+        }
+        catch (Exception)
+        {
+            return null;
+        }
+    }
+
+    public async Task<CmfPackageV1Collection> List()
+    {
+        // not all feeds support searching for keywords
+        var pkgs = await client.SearchPackages("keywords:cmf-deployment-package");
+        return new CmfPackageV1Collection();
+        // throw new System.NotImplementedException();
+    }
+
+    public async Task Put(CmfPackageV1 package)
+    {
+        var tmp = this.fileSystem.DirectoryInfo.New(this.fileSystem.Path.GetTempPath());
+        Log.Debug($"Downloading package {package.PackageAtRef} to {tmp.FullName}...");
+        var file = await package.Client.Get(package, tmp);
+        Log.Debug("Done!");
+        if (file.Extension == ".zip")
+        {
+            var zipFile = file;
+            file = this.fileSystem.FileInfo.New(this.fileSystem.Path.ChangeExtension(zipFile.FullName, "tgz"));
+            Log.Debug($"Package at {zipFile.FullName} is in Zip format, converting to tgz at {file.FullName}...");
+            CmfPackageController.ConvertZipToTarGz(zipFile, file);
+            Log.Debug("Done!");
+        }
+        else if (file.Extension == ".json")
+        {
+            // TODO: pack
+            throw new CliException("Please pack before publishing and publish the packed file.");
+        }
+        Log.Debug("Publishing via NPMClient...");
+        await this.client.PublishPackage(file);
+        Log.Debug("Publishing via NPMClient completed!");
+    }
+
+    public async Task<IFileInfo> Get(CmfPackageV1 package, IDirectoryInfo targetDirectory)
+    {
+        var tmp = targetDirectory.FileSystem.Path.GetTempFileName().Replace(".tmp", ".tgz");
```

Comment:
I don't know if this is still needed, but isn't it good practice to wrap the code in a try / finally to cleanup the temporary files?

Reply:
This is a good point and I'll add it to the cleanup issue.